Source/Core/DiscIO/VolumeWii.cpp

```diff
@@ -521,4 +527,143 @@ bool VolumeWii::CheckBlockIntegrity(u64 block_index, const Partition& partition)
+  bool error_occurred = false;
+
+  for (size_t i = 0; i < BLOCKS_PER_GROUP; ++i)
+  {
+    if (!error_occurred)
+    {
+      if (offset + (i + 1) * BLOCK_DATA_SIZE <= partition_data_decrypted_size)
+      {
+        if (!blob->ReadWiiDecrypted(offset + i * BLOCK_DATA_SIZE, BLOCK_DATA_SIZE,
+                                    unencrypted_data[i].data(), partition_data_offset))
+        {
+          error_occurred = true;
+        }
+      }
+      else
+      {
+        unencrypted_data[i].fill(0);
+      }
+    }
+
+    hash_futures[i] = std::async(std::launch::async, [&unencrypted_data, &unencrypted_hashes,
+                                                      &hash_futures, error_occurred, i]() {
+      const size_t h1_base = Common::AlignDown(i, 8);
+
+      if (!error_occurred)
+      {
+        // H0 hashes
+        for (size_t j = 0; j < 31; ++j)
+        {
+          mbedtls_sha1_ret(unencrypted_data[i].data() + j * 0x400, 0x400,
+                           unencrypted_hashes[i].h0[j]);
+        }
+
+        // H0 padding
+        std::memset(unencrypted_hashes[i].padding_0, 0, sizeof(HashBlock::padding_0));
+
+        // H1 hash
+        mbedtls_sha1_ret(reinterpret_cast<u8*>(unencrypted_hashes[i].h0), sizeof(HashBlock::h0),
+                         unencrypted_hashes[h1_base].h1[i - h1_base]);
+      }
+
+      if (i % 8 == 7)
+      {
+        for (size_t j = 0; j < 7; ++j)
+          hash_futures[h1_base + j].get();
+
+        if (!error_occurred)
+        {
+          // H1 padding
+          std::memset(unencrypted_hashes[h1_base].padding_1, 0, sizeof(HashBlock::padding_1));
+
+          // H1 copies
+          for (size_t j = 1; j < 8; ++j)
+          {
+            std::memcpy(unencrypted_hashes[h1_base + j].h1, unencrypted_hashes[h1_base].h1,
+                        sizeof(HashBlock::h1));
+          }
+
+          // H2 hash
+          mbedtls_sha1_ret(reinterpret_cast<u8*>(unencrypted_hashes[i].h1), sizeof(HashBlock::h1),
+                           unencrypted_hashes[0].h2[h1_base / 8]);
+        }
+
+        if (i == BLOCKS_PER_GROUP - 1)
+        {
+          for (size_t j = 0; j < 7; ++j)
+            hash_futures[j * 8 + 7].get();
+
+          if (!error_occurred)
+          {
+            // H2 padding
+            std::memset(unencrypted_hashes[0].padding_2, 0, sizeof(HashBlock::padding_2));
+
+            // H2 copies
+            for (size_t j = 1; j < BLOCKS_PER_GROUP; ++j)
+            {
+              std::memcpy(unencrypted_hashes[j].h2, unencrypted_hashes[0].h2,
+                          sizeof(HashBlock::h2));
+            }
+          }
+        }
+      }
+    });
+  }
+
+  // Wait for all the async tasks to finish
+  hash_futures.back().get();
+
+  if (error_occurred)
+    return false;
+
+  const size_t hardware_threads = std::thread::hardware_concurrency();
+  size_t threads = 1;
+  while (hardware_threads > threads && threads < BLOCKS_PER_GROUP)
```

Comment:
hardware_threads does not have to be a power of two, so it is better to check for hardware_threads >= 2*threads

Reply:
Just to make sure that I'm understanding this correctly, is the effect of your suggestion that we would use 4 threads on a 6-core CPU instead of 8? If so, I don't see why that's better.

Reply:
Both is as bad, indeed. I assume those tasks won't take more than 16ms, so 6 threads may finish before the other two have started. Why do you need this power of two constraint at all? Begin and end can be calcuated without a power of twp threads.

Reply:
i * BLOCKS_PER_GROUP / threads .. (i + 1) * BLOCKS_PER_GROUP / threads

Reply:
I suppose I did it because I figured the calculations needed for calculating start and end would be more complicated than that. I'll change it to not require a power of two.